ansible_base/lib/abstract_models/common.py

```diff
@@ -74,40 +75,69 @@ class Meta:
             if system_username is not None:
                 try:
                     user = get_user_model().objects.get(username=system_username)
                 except get_user_model().DoesNotExist:
-                    if warn_nonexistent_system_user:
-                        logger.warn(f"SYSTEM_USERNAME is set to {system_username} but no user with that username exists. User attribution will be None.")
+                    if non_existent_user_fatal:
+                        logger.warning(f"SYSTEM_USERNAME is set to {system_username} but no user with that username exists.")
                     user = None
+
         return user
 
-    def save(self, *args, warn_nonexistent_system_user=True, **kwargs):
+    def _check_user(self, user, non_existent_user_fatal=True):
+        # Its possible for _attributable_user to return None or anonymous user in which case we don't want to save
+        if (not user or user.is_anonymous) and non_existent_user_fatal:
+            # TODO: See if there is a better way to figure out how to identify this object?
+            # Maybe instead of trying to send a single identifier we try and dump the object?
+            obj_id = getattr(self, 'pk', getattr(self, 'name', getattr(self, 'username', 'Unknown')))
+            logger.error(f"Request to save a {self.__class__} object, id {obj_id} without a user making the request!")
+            raise ValueError("Unable to save model without user!")
+
+    def save(self, *args, non_existent_user_fatal=True, **kwargs):
         update_fields = list(kwargs.get('update_fields', []))
-        user = self._attributable_user(warn_nonexistent_system_user)
 
         # Manually perform auto_now_add and auto_now logic.
         now = timezone.now()
-        if not self.pk and not self.created_on:
-            self.created_on = now
-            if user is not None and not user.is_anonymous:
+        user = self._attributable_user(non_existent_user_fatal)
+
+        if not self.pk:
+            if self.created_by is None:
+                self._check_user(user, non_existent_user_fatal)
                 self.created_by = user
-            if 'created_on' not in update_fields:
-                update_fields.append('created_on')
-            if 'created_by' not in update_fields:
                 update_fields.append('created_by')
-        if 'modified_on' not in update_fields or not self.modified_on:
+
+            if self.created_on is None:
+                self.created_on = now
+                update_fields.append('created_on')
+
+        if 'modified_on' not in update_fields:
             self.modified_on = now
-            if user is not None and not user.is_anonymous:
-                self.modified_by = user
             update_fields.append('modified_on')
+
+        if 'modified_by' not in update_fields:
+            self._check_user(user, non_existent_user_fatal)
+            self.modified_by = user
             update_fields.append('modified_by')
```

Comment:
Might be good to document in a comment or method docstring that adding a field like "modified_by" to update_fields will be taken as a signal that the saver is providing the value. This is quite unlike the field `auto_now` type functions of datetime fields.

Reply:
Docstring added.